src/PreallocatedLinearOperators.jl

```diff
@@ -0,0 +1,135 @@
+export PreallocatedLinearOperator
+
+abstract type PreallocatedAbstractLinearOperator{T,F1,F2,F3} <: AbstractLinearOperator{T,F1,F2,F3} end
+
+"""
+Type to represent a linear operator with preallocation. Implicit modifications may
+happen if used without care.
+"""
+mutable struct PreallocatedLinearOperator{T,F1<:FuncOrNothing,F2<:FuncOrNothing,F3<:FuncOrNothing} <: PreallocatedAbstractLinearOperator{T,F1,F2,F3}
+  nrow   :: Int
+  ncol   :: Int
+  symmetric :: Bool
+  hermitian :: Bool
+  prod   :: F1 # apply the operator to a vector
+  tprod  :: F2 # apply the transpose operator to a vector
+  ctprod :: F3 # apply the transpose conjugate operator to a vector
+end
+
+"""
+    show(io, op)
+
+Display basic information about a linear operator.
+"""
+function show(io :: IO, op :: PreallocatedAbstractLinearOperator)
+  s  = "Preallocated linear operator\n"
+  s *= @sprintf("  nrow: %s\n", op.nrow)
+  s *= @sprintf("  ncol: %d\n", op.ncol)
+  s *= @sprintf("  eltype: %s\n", eltype(op))
+  s *= @sprintf("  symmetric: %s\n", op.symmetric)
+  s *= @sprintf("  hermitian: %s\n", op.hermitian)
+  s *= "\n"
+  print(io, s)
+end
+
+# Constructors.
+"""
+    PreallocatedLinearOperator(Mv, Mtu, Maw, M; symmetric=false, hermitian=false)
+
+Construct a linear operator from a dense or sparse matrix, using `Mv` as storage space
+for `M * v` and `Mtu` as storage space for `transpose(M) * u` and `Maw` to store
+`adjoint(M) * w`. Use the optional keyword arguments to indicate whether the operator
+is symmetric and/or hermitian.
+"""
+function PreallocatedLinearOperator(Mv :: Vector{T}, Mtu :: Vector{T}, Maw :: Vector{T},
+                                    M :: AbstractMatrix{T};
+                                    symmetric=false, hermitian=false) where T
+  nrow, ncol = size(M)
+  @assert length(Mv) ≥ nrow
+  @assert length(Mtu) ≥ ncol
+  @assert length(Maw) ≥ ncol
+  prod = @closure v -> mul!(Mv, M, v)
+  tprod = @closure u -> mul!(Mtu, transpose(M), u)
+  ctprod = @closure w -> mul!(Maw, adjoint(M), w)
+  F1 = typeof(prod)
+  F2 = typeof(tprod)
+  F3 = typeof(ctprod)
+  PreallocatedLinearOperator{T,F1,F2,F3}(nrow, ncol, symmetric, hermitian, prod, tprod, ctprod)
+end
+
+"""
+    PreallocatedLinearOperator(Mv, M; symmetric=false, hermitian=false)
+
+Construct a linear operator with preallocation using `Mv` as storage space for all
+matrix-vector products.
```

Comment:
I'm not sure I understand. In the function you allocate storage for the other types of product. What's the point of this function? Do you mean that you will allocate the minimum amount of storage required depending on whether M is symmetric and/or hermitian?

Reply:
This is a shorter API to use the same memory for all three products. Assuming the user has use for it, this is simply shorter to write.

Reply:
Ok. I think the docstring should contain appropriate warnings that `Mv` will be overwritten by each type of operation. Also this only makes sense for square M?!

Reply:
I suppose the `@assert`s in the previous function will catch dimension issues.

Reply:
Actually, I was under the impression that `mul!(Av, A, v)` worked if `length(Av) > size(A, 1)`, but it doesn't.
Using `mul!(Av[1:nrow], A, v)` but it allocates more and is slower, so I think I have to enforce dimension equality.